### PR TITLE
Introduce persistent and explicit locking of file and folders

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Description
Introduce persistent locks to owncloud.

since we need to list all locks in sub trees I propose to implement these persistent locks in database only and use a join with the file cache.

structure for oc_persistent_file-locks

Column Name | Type | Description
------------ | ------------- | -------------
id | int | PK and auto increment
file_id | int not null| FK to fileid in oc_file_cache
owner | varchar | displayname of locking user
timeout | int | eol timestamp of the lock
created_at | int | timestamp lock was created
token | varchar(1024) | uuid for webdav - 1024 chars for wopi
scope | int | 1 = exclusive, 2 = shared
depth | int | 0, 1 or infinity

recursive query will look like (needed on DELETE and MOVE - any other request only need lock information about the resource itself)
```sql
select l.*, f.path
from oc_lock l
join oc_file_cache f on l.file_id = f.fileid
where f.path like 'files/user/foo/%'
```

## ToDo in this PR
- [x] implement locking in db provider
- [x] implement a real webdav locking provider

## ToDo in follow up PRs or othre repos
- [x] add locking to wopi app => https://github.com/owncloud/wopi/issues/6
- [x] add file action in web ui to lock, unlock files and folders
- [x] display lock information in web ui
- [ ] coordinate behavior with mobile and desktop client
- [ ] cleanup process of outdated locks
- [x] litmus on shared folder

## Motivation and Context
- required by some integrations - wopi
- required in some use cases - e.g. working on big file over a longer period of time

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by PVince81** [comment]
conflicts found

**Comment by PVince81** [comment]
As discussed before, for litmus one idea would be to have a behat scenario step "and Webdav path /XYZ is Webdav compliant" that runs litmus in the background. This way you could pass any path like external storages, shared folders, etc to make sure all Webdav operations work correctly there.

does that make sense or is this too far-fetched ? @patrickjahns @phil-davis

File: apps/dav/tests/unit/Files/FileLocksBackendTest.php [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+class FileLocksBackendTest extends TestCase {
+
+	/** @var FileLocksBackend */
+	private $plugin;
+	/** @var Tree | \PHPUnit_Framework_MockObject_MockObject */
+	private $tree;
+	/** @var IPersistentLockingStorage | IStorage | \PHPUnit_Framework_MockObject_MockObject */
+	private $storageOfFileToBeLocked;
+
+	public function setUp() {
```
Comment (PVince81) [inline apps/dav/tests/unit/Files/FileLocksBackendTest.php]: also add tests for the `before` part of the plugin, as its logic is non-trivial

File: apps/files/appinfo/info.xml [hunk 2]
```diff
@@ -6,7 +6,7 @@
 	<licence>AGPL</licence>
 	<author>Robin Appelman, Vincent Petry</author>
 	<default_enable/>
-	<version>1.5.1</version>
+	<version>1.5.2</version>
```
Comment (PVince81) [inline apps/files/appinfo/info.xml]: maybe 1.6.0, that's a big addition like a new feature :-)

File: apps/files/js/filelockplugin.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+		},
+
+		/**
+		 * @param fileList
+		 * @private
+		 */
+		_extendFileActions: function(fileList) {
+			var self = this;
+			// fileActions.registerAction({
```
Comment (PVince81) [inline apps/files/js/filelockplugin.js]: please remove these. that could be a separate app

File: apps/files/js/locktabview.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+			events: {
+				'click a.unlock': '_onClickUnlock'
+			},
+
+			_onClickUnlock: function (event) {
+				var self = this;
+				var lockToken = event.target.getAttribute('data-lock-token');
+				var lockUrl = event.target.getAttribute('data-lock-root');
+
+				this.model._filesClient.getClient().request('UNLOCK',
```
Comment (PVince81) [inline apps/files/js/locktabview.js]: if this becomes common, we could add `lock(path, token)` and `unlock(path, token)` as direct methods on the Files.Client (and/or js-owncloud-client). I could picture apps like the text editor using this method to lock the file while the editor is still open

File: apps/files/js/locktabview.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+					lockUrl,
+					{
+						'Lock-Token': lockToken
+					}).then(function (result) {
+						if (result.status === 204) {
+							var locks = self.model.get('activeLocks');
+							locks = locks.filter(function(item) {
+								return item.locktoken !== lockToken;
```
Comment (PVince81) [inline apps/files/js/locktabview.js]: does it mean the server might return lock tokens not related to the one you passed in the header, and you need to filter them out ? if yes, add as code comment to clarify

File: apps/files/lib/BackgroundJob/CleanupPersistentFileLocks.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace OCA\Files\BackgroundJob;
+
+use OC\BackgroundJob\TimedJob;
+use OC\Lock\Persistent\LockMapper;
+
+/**
+ * Clean up persistent file locks that are expired
+ */
+class CleanupPersistentFileLocks extends TimedJob {
```
Comment (PVince81) [inline apps/files/lib/BackgroundJob/CleanupPersistentFileLocks.php]: aha, I think you can then tick the checkbox about lock cleanup, unless it's something else ?

File: apps/files_sharing/lib/SharedStorage.php [hunk 7]
```diff
@@ -456,4 +457,23 @@ public function getWrapperStorage() {
 
 		return $this->sourceStorage;
 	}
+
+	public function getLocks(string $internalPath, bool $returnChildLocks = false): array {
+		$locks = parent::getLocks($this->getSourcePath($internalPath), $returnChildLocks);
+		return \array_map(function (ILock $lock) {
+			// TODO: if path starts with rootpath
```
Comment (PVince81) [inline apps/files_sharing/lib/SharedStorage.php]: is that something to be concerned about before merging ?

File: lib/private/Files/Storage/Common.php [hunk 8]
```diff
@@ -710,15 +713,24 @@ public function getVersions($internalPath) {
...
-	protected function convertInternalPathToGlobalPath($internalPath) {
+	public function convertInternalPathToGlobalPath($internalPath) {
 		$mounts = \OC::$server->getMountManager()->findByStorageId($this->getId());
-		$mount = \end($mounts);
-		$p = $mount->getMountPoint() . $internalPath;
+
+		$selectedMount = \end($mounts);
+		foreach ($mounts as $mount) {
+			$o = \explode('/', $mount->getMountPoint());
+			if ($o[1] === $this->owner) {
```
Comment (PVince81) [inline lib/private/Files/Storage/Common.php]: hope this works with public links when shared from a system-wide external storage, to be verified

File: lib/private/Files/Storage/Common.php [hunk 9]
```diff
@@ -748,4 +760,34 @@ public function saveVersion($internalPath) {
...
+		$storageId = $this->getCache()->getNumericStorageId();
+		$fileId = $this->getCache()->getId($internalPath);
+		return $locksManager->lock($storageId, $internalPath, $fileId, $lockInfo);
+	}
+
+	public function unlockNodePersistent(string $internalPath, array $lockInfo) {
+		/** @var LockManager $locksManager */
+		$locksManager = \OC::$server->query(LockManager::class);
+		$fileId = $this->getCache()->getId($internalPath);
+		return $locksManager->unlock($fileId, $lockInfo['token']);
```
Comment (PVince81) [inline lib/private/Files/Storage/Common.php]: hmm, so lock() needs storageId, internalPath and $fileId but unlock only needs $fileId ?
why is $fileid alone not enough for `lock()` ?

File: lib/private/Files/Storage/Common.php [hunk 10]
```diff
@@ -748,4 +760,34 @@ public function saveVersion($internalPath) {
...
+		/** @var LockManager $locksManager */
+		$locksManager = \OC::$server->query(LockManager::class);
+		$storageId = $this->getCache()->getNumericStorageId();
+		$fileId = $this->getCache()->getId($internalPath);
+		return $locksManager->lock($storageId, $internalPath, $fileId, $lockInfo);
+	}
+
+	public function unlockNodePersistent(string $internalPath, array $lockInfo) {
+		/** @var LockManager $locksManager */
+		$locksManager = \OC::$server->query(LockManager::class);
```
Comment (PVince81) [inline lib/private/Files/Storage/Common.php]: expose with convenience method on server container?

File: lib/private/Lock/Persistent/Lock.php [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+	protected $depth;
+	/** @var int - foreign key zu oc_account.id */
+	protected $ownerAccountId;
+
+	/** @var string - joined with oc_filecache */
+	protected $path;
+	/** @var string - computed value */
+	protected $globalUserId;
+	/** @var string - computed value */
+	protected $globalFileName;
```
Comment (PVince81) [inline lib/private/Lock/Persistent/Lock.php]: what's a "globalFileName" and a "globalUserId" ? explain in the comment

maybe rename globalFileName to absolutePath

File: lib/private/Lock/Persistent/LockManager.php [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+		}
+
+		$locks = $this->lockMapper->getLocksByPath($storageId, $internalPath, false);
+		$exists = false;
+		foreach ($locks as $lock) {
+			if ($lock->getToken() === $lockInfo['token']) {
+				$exists = true;
+				$lock->setCreatedAt(\time());
+				$lock->setTimeout($timeout);
+				if (empty($lock->getOwner())) {
```
Comment (PVince81) [inline lib/private/Lock/Persistent/LockManager.php]: hopefully the owner will never be called "0" ?

File: lib/private/Lock/Persistent/LockManager.php [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+			$depth = $lockInfo['depth'];
+		}
+		$scope = ILock::LOCK_SCOPE_EXCLUSIVE;
+		if (isset($lockInfo['scope'])) {
+			$scope = $lockInfo['scope'];
+		}
+
+		$lock = new Lock();
+		$lock->setFileId($fileId);
+		$lock->setCreatedAt(\time());
```
Comment (PVince81) [inline lib/private/Lock/Persistent/LockManager.php]: use time factory for testability
also add unit tests ;-)

File: lib/private/Lock/Persistent/LockMapper.php [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+	 */
+	public function getLocksByPath(int $storageId, string $internalPath, bool $returnChildLocks) : array {
+		$query = $this->db->getQueryBuilder();
+		$pathPattern = $this->db->escapeLikeParameter($internalPath) . '%';
+
+		$query->select(['id', 'owner', 'timeout', 'created_at', 'token', 'token', 'scope', 'depth', 'file_id', 'path', 'owner_account_id'])
+			->from($this->getTableName(), 'l')
+			->join('l', 'filecache', 'f', $query->expr()->eq('l.file_id', 'f.fileid'))
+			->where($query->expr()->eq('storage', $query->createNamedParameter($storageId)))
+			->andWhere($query->expr()->gt('created_at', $query->createFunction('(' . $query->createNamedParameter(\time()) . ' - `timeout`)')));
```
Comment (PVince81) [inline lib/private/Lock/Persistent/LockMapper.php]: use time factory for testability

File: lib/private/Lock/Persistent/LockMapper.php [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+
+		$currentPath = '';
+		foreach ($uriParts as $part) {
+			if ($currentPath) {
+				$currentPath .= '/';
+			}
+			$currentPath .= $part;
+			$query->orWhere(
+				$query->expr()->andX(
+					// TODO: think about parent locks for depth 1
```
Comment (PVince81) [inline lib/private/Lock/Persistent/LockMapper.php]: should we meet and talk about these ?

File: lib/private/Lock/Persistent/LockMapper.php [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+			throw new \InvalidArgumentException('token_hash does not match the token of the lock');
+		}
+		return parent::update($entity);
+	}
+
+	public function cleanup() {
+		$query = $this->db->getQueryBuilder();
+		$query->delete($this->getTableName())
+			->where($query->expr()->lt('created_at',
+				$query->createFunction('(' . $query->createNamedParameter(\time()) . ' - `timeout`)')))
```
Comment (PVince81) [inline lib/private/Lock/Persistent/LockMapper.php]: time factory + unit test

**Review by PVince81 — COMMENTED** [review]
Reviewed.

There seems to be a few open topics (TODOs) to look at.

For the JS code we could move it to a separate PR and add unit tests there.

Also potentially missing tests in the sections about "use time factory".

**Comment by PVince81** [comment]
can you also update the top post to match the new data model ? seems there are some new fields like "global user id" and "global file name"

File: lib/private/Files/Storage/Common.php (hunk 8 above)
Comment (DeepDiver1975) [reply lib/private/Files/Storage/Common.php]: To be tested ....

File: lib/private/Files/Storage/Common.php (hunk 10 above)
Comment (DeepDiver1975) [reply lib/private/Files/Storage/Common.php]: I don't think this is needed in Public API level

File: lib/private/Lock/Persistent/Lock.php (hunk 11 above)
Comment (DeepDiver1975) [reply lib/private/Lock/Persistent/Lock.php]: The path via webdav. E.g. Foo/bar.txt

Name this absolute path? Sounds good

File: lib/private/Files/Storage/Common.php (hunk 9 above)
Comment (DeepDiver1975) [reply lib/private/Files/Storage/Common.php]: I need to Look into the implementation

File: apps/dav/lib/Connector/Sabre/LockPlugin.php [hunk 17]
```diff
@@ -81,4 +86,26 @@ public function releaseLock(RequestInterface $request) {
...
+	 */
+	public function beforeUnlock($uri, LockInfo $lock) {
+		/** @var LockMapper $mapper */
+		$mapper = \OC::$server->query(LockMapper::class);
+
+		$lock = $mapper->getLockByToken($lock->token);
+		if ($lock === null || $lock->getOwnerAccountId() === null) {
+			return;
+		}
+		$currentUser = \OC::$server->getUserSession()->getUser();
```
[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply apps/dav/lib/Connector/Sabre/LockPlugin.php]: Test coverage needs to be pushed. Will work ob this asap

File: apps/dav/lib/Connector/Sabre/LockPlugin.php [hunk 18]
```diff
@@ -81,4 +86,26 @@ public function releaseLock(RequestInterface $request) {
...
+	 * @param $uri
+	 * @param LockInfo $lock
+	 * @throws Forbidden
+	 * @throws \OCP\AppFramework\Db\DoesNotExistException
+	 * @throws \OCP\AppFramework\Db\MultipleObjectsReturnedException
+	 * @throws \OCP\AppFramework\QueryException
+	 */
+	public function beforeUnlock($uri, LockInfo $lock) {
+		/** @var LockMapper $mapper */
+		$mapper = \OC::$server->query(LockMapper::class);
```
[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply apps/dav/lib/Connector/Sabre/LockPlugin.php]: comment added to code

File: apps/dav/lib/Files/FileLocksBackend.php [hunk 19]
```diff
@@ -75,12 +75,12 @@ public function getLocks($uri, $returnChildLocks) {
 		$davLocks = [];
 		foreach ($locks as $lock) {
 			$lockInfo = new Locks\LockInfo();
-			$fileName = $lock->getGlobalFileName();
+			$fileName = $lock->getAbsolteDavPath();
```
Comment (phil-davis) [inline apps/dav/lib/Files/FileLocksBackend.php]: It is going to explode:
```
public function getAbsoluteDavPath()
````

s/getAbsolteDavPath/getAbsoluteDavPath

**Comment by DeepDiver1975** [comment]
> can you also update the top post to match the new data model ? seems there are some new fields like "global user id" and "global file name"

these two fields are not in the database - thats why I did not add them above

File: apps/dav/lib/Connector/Sabre/LockPlugin.php (hunk 17 above)
[the remark this answers is not in the document]
Comment (PVince81) [reply apps/dav/lib/Connector/Sabre/LockPlugin.php]: did you end up injecting or did you decide against it ?

File: apps/dav/tests/unit/DAV/LockPluginTest.php [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+		$this->server = $this->createMock(Server::class);
+		$this->tree = $this->createMock(Tree::class);
+		$this->server->tree = $this->tree;
+
+		$this->plugin->initialize($this->server);
+		$this->lockMapper = $this->createMock(LockMapper::class);
+		$this->userSession = $this->createMock(IUserSession::class);
+
+		$this->overwriteService(LockMapper::class, $this->lockMapper);
+		$this->overwriteService('UserSession', $this->userSession);
```
Comment (PVince81) [inline apps/dav/tests/unit/DAV/LockPluginTest.php]: hmm ok, so you overrode it

File: apps/dav/tests/unit/Files/FileLocksBackendTest.php (hunk 1 above)
Comment (PVince81) [reply apps/dav/tests/unit/Files/FileLocksBackendTest.php]: ok, seems to be done

File: apps/files/appinfo/info.xml (hunk 2 above)
Comment (PVince81) [reply apps/files/appinfo/info.xml]: @DeepDiver1975 ping

File: apps/dav/lib/Connector/Sabre/LockPlugin.php (hunk 17 above)
[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply apps/dav/lib/Connector/Sabre/LockPlugin.php]: decided against for perf reasons

File: apps/dav/tests/unit/Files/FileLocksBackendTest.php (hunk 1 above)
Comment (DeepDiver1975) [reply apps/dav/tests/unit/Files/FileLocksBackendTest.php]: done

File: apps/dav/lib/Files/FileLocksBackend.php [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+			return [];
+		}
+
+		/** @var IPersistentLockingStorage $storage */
+		$locks = $storage->getLocks($node->getFileInfo()->getInternalPath(), $returnChildLocks);
+
+		$davLocks = [];
+		foreach ($locks as $lock) {
+			$lockInfo = new Locks\LockInfo();
+			$fileName = $lock->getGlobalFileName();
```
[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply apps/dav/lib/Files/FileLocksBackend.php]: done

File: apps/dav/lib/Files/FileLocksBackend.php [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+
+			if ($this->useV1) {
+				$lockInfo->uri = $fileName;
+			} else {
+				$uid = $lock->getGlobalUserId();
```
[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply apps/dav/lib/Files/FileLocksBackend.php]: done - dav user as discussed

File: apps/files/appinfo/info.xml (hunk 2 above)
Comment (DeepDiver1975) [reply apps/files/appinfo/info.xml]: moved to own pr

File: apps/files/js/filelockplugin.js (hunk 3 above)
Comment (DeepDiver1975) [reply apps/files/js/filelockplugin.js]: own pr

File: apps/files/js/locktabview.js (hunk 4 above)
Comment (DeepDiver1975) [reply apps/files/js/locktabview.js]: if we need one day - yes - but anyway own pr

File: apps/files/js/locktabview.js (hunk 5 above)
Comment (DeepDiver1975) [reply apps/files/js/locktabview.js]: own pr - will add explanation there

File: apps/files/lib/BackgroundJob/CleanupPersistentFileLocks.php (hunk 6 above)
Comment (DeepDiver1975) [reply apps/files/lib/BackgroundJob/CleanupPersistentFileLocks.php]: yes - ticked long time back

File: apps/files_sharing/lib/SharedStorage.php (hunk 7 above)
Comment (DeepDiver1975) [reply apps/files_sharing/lib/SharedStorage.php]: can be removed - old comment

File: lib/private/Files/Storage/Common.php (hunk 9 above)
Comment (DeepDiver1975) [reply lib/private/Files/Storage/Common.php]: the lock operation needs information about path and storage due to it's implementation

File: lib/private/Lock/Persistent/LockManager.php (hunk 12 above)
Comment (DeepDiver1975) [reply lib/private/Lock/Persistent/LockManager.php]: alternative !== '' && !== null :question:

Comment (PVince81) [reply lib/private/Lock/Persistent/LockManager.php]: yeah, that's what we usually do

**Comment by individual-it** [comment]
> As discussed before, for litmus one idea would be to have a behat scenario step "and Webdav path /XYZ is Webdav compliant" that runs litmus in the background.

Actually we only need a `Then` line but to make it more compliant with Behat structure I'm thinking along this lines
```
When the path "/XYZ" is used as webDav path
Then all litmus checks should pass
```
The challenge would be to change the way the litmus docker container works. Currently we set the URL, start the container and the container checks the URL. We would need to change the container, so that it works as an service with some kind of simple API. We would give it an URL, it runs the checks, returns the result and waits for the next request.

@PVince81 @DeepDiver1975 @tboerger @patrickjahns @phil-davis

**Comment by DeepDiver1975** [comment]
> The challenge would be to change the way the litmus docker container works. Currently we set the URL, start the container and the container checks the URL. We would need to change the container, so that it works as an service with some kind of simple API. We would give it an URL, it runs the checks, returns the result and waits for the next request.

can't we do an exec in php of the command: docker run -e LITMUS_URL ...... litmus:latest and parse the output?

Building a service with API and anything related is a bit time consuming

**Comment by individual-it** [comment]
I don't think we can start a new docker from within a drone run, because when we are running PHP we are already in a docker containers.
We only need a service listening on a port, receiving 3 variables and starting a command. Maybe a bit of curl magic would do

**Comment by patrickjahns** [comment]
Can you please clarify, why you need to combine behat and litmus? Why not simply use either one for this?

**Comment by individual-it** [comment]
@PVince81 @DeepDiver1975 ^^

**Comment by DeepDiver1975** [comment]
This was the Idea. https://github.com/owncloud/core/pull/31651#issuecomment-409601585

**Comment by patrickjahns** [comment]
I can read that - but why does it need to be combined - why can't we just setup the parts beforehand and then run the litmus container in a seperate step with the previously setup path?

Alternatively - port the litmus tests into a behat suite an we provide it as package for others to use again

**Comment by DeepDiver1975** [comment]
> I can read that - but why does it need to be combined - why can't we just setup the parts beforehand and then run the litmus container in a seperate step with the previously setup path?

the idea was to reuse the functionalities we have in the behat env like creating a local mounted folder, a shared folder and so on and run litmus then on these folders.

If we cannot combine both we need to reinvent this for our litmus test suite



> Alternatively - port the litmus tests into a behat suite an we provide it as package for others to use again

I think this is out of scope .... but worth a thought ...

**Comment by individual-it** [comment]
> Building a service with API and anything related is a bit time consuming

what about having a php file inside the litmus container run by php-dev server just listening and executing litmus with parameters whenever there is a request

**Comment by PVince81** [comment]
assigning self for re-review

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by DeepDiver1975** [comment]
cool - thx

**Comment by dalenoe** [comment]
Will this work with redis and/or memcached?

**Comment by DeepDiver1975** [comment]
> Will this work with redis and/or memcached?

database

**Pull request merged**